### DISCUSSION TITLE
clickhouse udf  wrong result, expected 1 row(s), actual 0

### DISCUSSION THREAD

**Discussion by QiMingChina** [ClickHouse/ClickHouse#53262] [Q&A]
I use Python to write clickhouse udf
clickhouse version 23.3.2 Python 3.9.12
I added the configuration <user_defined_executable_functions_config>*_function.xml</user_defined_executable_functions_config> in the config.xml file
I added the configuration file  getusername_function.xml and added the clickhouse user permission in the/etc/clickhouse server path
![image](https://github.com/ClickHouse/ClickHouse/assets/48780003/a8d47f35-b8e9-4ad0-ab47-418a9856b292)

```
<functions>
    <function>
        <type>executable_pool</type>
        <name>username_function</name>
        <return_type>String</return_type>
        <return_name>result_username</return_name>
        <argument>
            <type>String</type>
            <name>first_name</name>
        </argument>
        <argument>
            <type>String</type>
            <name>second_name</name>
        </argument>
        <argument>
            <type>String</type>
            <name>hobby</name>
        </argument>
        <format>JSONEachRow</format>
        <command>username.py</command>
    </function>
</functions>
```
I  place Python file a in user_scripts path 
![image](https://github.com/ClickHouse/ClickHouse/assets/48780003/268c8767-af56-4bae-8b54-b4fe7e0b1cbd)

```
#!/servers/software/anaconda3/bin/python

import sys
import json
import logging

logging.basicConfig()
logging.getLogger().setLevel(logging.INFO)
file_handler = logging.FileHandler('/data/clickHousedata/user_scripts/username-log.txt')
logging.getLogger().addHandler(file_handler)


def username(first_name, second_name, hobby):
    user_info = 'NAME:{0}  HOBBY:{1}'.format(first_name + second_name, hobby)
    return user_info


if __name__ == '__main__':
    for line in sys.stdin:
        logging.info('输入参数列表：{0}'.format(line))
        arg_dict = json.loads(line)
        first_name = arg_dict['first_name']
        second_name = arg_dict['second_name']
        hobby = arg_dict['hobby']
        info=username(first_name,second_name,hobby)
        result = {'result_username':info}
        print(json.dumps(result), end='\n')
        sys.stdout.flush()
```

reload function
![image](https://github.com/ClickHouse/ClickHouse/assets/48780003/6d12cd48-a513-4a6d-9851-35451f15d526)

run script
```
echo '{ "first_name": "first", "second_name": "second", "hobby": "hobby" }' | ./username.py
```
![image](https://github.com/ClickHouse/ClickHouse/assets/48780003/aaa0976b-1f50-400a-91d2-1716089be87a)

raise error
![image](https://github.com/ClickHouse/ClickHouse/assets/48780003/5256c9ba-3832-4e4c-8d77-7710097bd02e)
````
 Function 'username_function': wrong result, expected 1 row(s), actual 0: While processing username_function('a', 'd', 'football'). (UNSUPPORTED_METHOD)
```

**Comment by QiMingChina** [comment]
Could you help me take a look? @den-crane @alexey-milovidov @alesapin

**Comment by krutik228** [comment]
Same problem:
```bash
/var/lib/clickhouse/user_scripts # ls -lah
-rwxrwxrwx    1 root     root         786 Aug 14 08:22 v1_greater_or_equal_v2.py
```
/etc/clickhouse-server/v1_greater_or_equal_v2_function.xml
```xml
<functions>
    <function>
        <type>executable</type>
        <name>v1_greater_or_equal_v2</name>
        <return_type>UInt64</return_type>
        <argument>
            <type>String</type>
            <name>version_1</name>
        </argument>
        <argument>
            <type>String</type>
            <name>version_2</name>
        </argument>
        <command>v1_greater_or_equal_v2.py</command>
        <format>TabSeparated</format>
        <execute_direct>1</execute_direct>
    </function>
</functions>
```

/var/lib/clickhouse/user_scripts/v1_greater_or_equal_v2.py

```python
#!/usr/bin/python3

import sys
import json


def normalize(version):
    return [int(v) for v in version.split(".")]


def v1_greater_or_equal_v2(v1, v2):
    v1_parts = normalize(v1)
    v2_parts = normalize(v2)

    max_length = max(len(v1_parts), len(v2_parts))

    v1_parts.extend([0] * (max_length - len(v1_parts)))
    v2_parts.extend([0] * (max_length - len(v2_parts)))

    for num1, num2 in zip(v1_parts, v2_parts):
        if num1 > num2:
            return 1
        elif num1 < num2:
            return 0

    return 1


if __name__ == '__main__':
    for line in sys.stdin:
        value = json.loads(line)
        v1 = value['version_1']
        v2 = value['version_2']

        res_int = v1_greater_or_equal_v2(v1, v2)

        print(res_int, end='')
        sys.stdout.flush()

```

Clickhouse error:
SQL Error [1] [07000]: Code: 1. DB::Exception: Function 'v1_greater_or_equal_v2': wrong result, expected 1 row(s), actual 0: While processing v1_greater_or_equal_v2('3.2', '5.2'). (UNSUPPORTED_METHOD) (version 22.6.9.11 (official build))
, server ClickHouseNode [uri=http://localhost:8123/default, options={use_server_time_zone=false,use_time_zone=false}]@1712656529

**Reply by QiMingChina** [reply]
```
<functions>
    <function>
        <type>executable</type>
        <name>v1_greater_or_equal_v2</name>
        <return_type>UInt64</return_type>
        <return_name>res</return_name>
        <argument>
            <type>String</type>
            <name>version_1</name>
        </argument>
        <argument>
            <type>String</type>
            <name>version_2</name>
        </argument>
        <command>v1_greater_or_equal_v2.py</command>
        <format>JSONEachRow</format>
    </function>
</functions>
```

```
if __name__ == '__main__':
    for line in sys.stdin:
        value = json.loads(line)
        v1 = value['version_1']
        v2 = value['version_2']

        res_int = v1_greater_or_equal_v2(v1, v2)
        result = {'res':res_int}
        print(json.dumps(result),end='\n')
        sys.stdout.flush()
```
Change the configuration file and code to this and try again

**Comment by lingithublearn** [comment]
i meet the same problem 
```  
 <function>
        <type>executable</type>
        <name>FUNC_业务_LTE_干扰_是否有效PRB</name>
        <return_type>UInt8</return_type>
        <argument>
            <type>Float64</type>
        </argument>
        <format>TabSeparated</format>
        <execute_direct>0</execute_direct>
        <command>python3 /mastercom/ck/data/clickhouse/user_scripts/FUNC_业务_LTE_干扰_是否有效PRB.py</command>
    </function>
```
```
#!/usr/bin/python3
import sys

if __name__ == '__main__':
	for prb in sys.stdin:
		result = 0
		prb = float(prb)
		if prb < 0 and prb >= -160:
			result = 1
		print(result)
		sys.stdout.flush()
```
select `FUNC_业务_LTE_干扰_是否有效PRB`(-10);
error
Received exception from server (version 22.6.7):
Code: 1. DB::Exception: Received from localhost:9000. DB::Exception: Function 'FUNC_业务_LTE_干扰_是否有效PRB': wrong result, expected 1 row(s), actual 0: While processing FUNC_业务_LTE_干扰_是否有效PRB(-10.). (UNSUPPORTED_METHOD)
0. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0xb8a3a9a in /usr/bin/clickhouse
1. DB::Exception::Exception<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, unsigned long&, unsigned long&>(int, fmt::v8::basic_format_string<char, fmt::v8::type_identity<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >::type, fmt::v8::type_identity<unsigned long&>::type, fmt::v8::type_identity<unsigned long&>::type>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&, unsigned long&, unsigned long&) @ 0x130b95ce in /usr/bin/clickhouse
2. DB::UserDefinedFunction::executeImpl(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&, std::__1::shared_ptr<DB::IDataType const> const&, unsigned long) const @ 0x1652f024 in /usr/bin/clickhouse
3. DB::IFunction::executeImplDryRun(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&, std::__1::shared_ptr<DB::IDataType const> const&, unsigned long) const @ 0xdf69aca in /usr/bin/clickhouse
4. DB::FunctionToExecutableFunctionAdaptor::executeDryRunImpl(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&, std::__1::shared_ptr<DB::IDataType const> const&, unsigned long) const @ 0xdf692ae in /usr/bin/clickhouse
5. DB::IExecutableFunction::executeWithoutLowCardinalityColumns(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&, std::__1::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x1537941d in /usr/bin/clickhouse
6. DB::IExecutableFunction::defaultImplementationForConstantArguments(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&, std::__1::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x15378e3a in /usr/bin/clickhouse
7. DB::IExecutableFunction::executeWithoutLowCardinalityColumns(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&, std::__1::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x153793c5 in /usr/bin/clickhouse
8. DB::IExecutableFunction::executeWithoutSparseColumns(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&, std::__1::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x15379e1a in /usr/bin/clickhouse
9. DB::IExecutableFunction::execute(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&, std::__1::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x1537b27c in /usr/bin/clickhouse
10. DB::ActionsDAG::addFunction(std::__1::shared_ptr<DB::IFunctionOverloadResolver> const&, std::__1::vector<DB::ActionsDAG::Node const*, std::__1::allocator<DB::ActionsDAG::Node const*> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) @ 0x15a04587 in /usr/bin/clickhouse

**Reply by lingithublearn** [reply]
i text change the foramat to JSONEachRow, it still show the same problem
```
<function>
        <type>executable</type>
        <name>json_text</name>
        <return_type>Float32</return_type>
        <return_name>result</return_name>
        <argument>
            <type>Float32</type>i
            <name>prb</name>
        </argument>
        <argument>
            <type>Float32</type>
            <name>band</name>
        </argument>
        <format>JSONEachRow</format>
        <execute_direct>0</execute_direct>
        <command>python3 /data/clickhouse/user_scripts/json_text.py</command>
    </function>
```
```
import sys
import json

line = sys.stdin.readline().strip()
value = json.loads(line)
avalue = int(value['prb'])
bvalue = int(value['band'])
num = bvalue
result = {"result": num}
print(json.dumps(result),end='\n')
sys.stdout.flush()
```
select json_text(2,2)
error
wrong result, expected 1 row(s), actual 0

![HW685JT 18W$R0VG_ %TM9](https://github.com/ClickHouse/ClickHouse/assets/87681054/10db40ae-e7fa-4117-b18d-dcb626f709fa)

the script can work
